### ISSUE TITLE
Buffer full while polling websocket

### ISSUE THREAD

**Issue by LikeLakers2** [godotengine/godot#22496]
**Godot version:**
v.3.1.alpha.custom_build. a2b6be2

**OS/device including version:**
Windows 10 Home v1803 (April 2018 update) x64

**Issue description:**
When polling the websocket and receiving a payload that in total would be bigger than 65535 bytes over all packets it takes to construct it, the following message prints out in the console:

```
ERROR: Buffer full! Dropping data
   At: modules\websocket\lws_peer.cpp:69
```

While I understand this seems to be intentional after digging into the code, this has two main issues:

1. It seems to corrupt the payload that the developer's code will get when checking for payloads. If I have a JSON payload where I receive, let's say `{"data":{"name":"Help","age":42}}`, imagine only getting `:{"name":"Help","age":42` -- that's basically what's happening here.

2. A websocket payload can be bigger than 65535 bytes, so why does it seem like this websocket implementation is limiting that size?

**Steps to reproduce:**
Receive a single websocket payload that would be more than 65535 bytes (including the 5-byte header) total.

**Minimal reproduction project:**
N/A (Without giving out a token my project uses for a service, I can't really provide a reproduction project, so you may have to set up a websocket server that will send out a payload that big)

**Comment by LikeLakers2** [comment]
(If you're viewed the original post through an email notification, please come view the latest version of it on Github. I've edited the original post a lot to better represent my understanding of the issue. I think the latest version is also a lot easier to understand in general.)

**Comment by LikeLakers2** [comment]
> Yes, although when the buffer is full, we have no choice but to drop packets. We might try to look back in the buffer and discard the whole message (not just the remaining frames), but that might still not be 100% effective.

Personally, I'd prefer dropping, though I don't know if other developers would.

> Additionally, this usually means you are sending too much data too fast, which will in any case result in other issues when you move to a real network instead of localhost/LAN.
> ...
> That said, the buffer was set to 64k at it is usually a good buffer size, especially in games, where you don't want to send a lot of data per second (i.e. you are not doing file transfers). And for the rare cases where you need to, you can always split it up in chunks yourself in your code.

Yeah, I would normally split it up, but as my use case for websockets involves an external service that I doubt would implement such an option just for me, I can't. And it's not like the external service is trying to send all of the payload at once -- it is properly splitting up the payload into many frames. It's just that, Godot's buffer is too small when trying to reconstruct the frames into a single payload. It wouldn't matter if I was on localhost/LAN or a real network, I would be encountering the 64k limitation either way.

> We could make it configurable, but it will need to be done prior to opening/connecting .
> And the buffer need, in our case, to be preallocated. So preallocating 2gb is not an option.

That'd be fine.

Also, 2GB? My use case would probably do well with just a ~10MB buffer, lol. I'm saving the rest of that memory for other things.

**Comment by mhilbrunner** [comment]
I think making the buffer size configurable is a reasonable option. :)

**Comment by LikeLakers2** [comment]
After messing around with the websocket code a bit, I've managed to figure out how to manually set the buffer size, and documented it on a commit in my own repo [here](https://github.com/LikeLakers2/godot/commit/81e3a4fbe3a5f1d160029eb28325de8535276040). That commit shows an example of changing the buffer from 64kb to 16mb. (yes, absurd example, I should never need to send a 16mb packet, but still)

I'd still like to be able to resize the buffer from GDScript without having to rebuild the engine, but for now this is a suitable alternative in my eyes.
